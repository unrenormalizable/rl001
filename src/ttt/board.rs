use serde::*;
use std::collections::HashMap;
use std::option::*;

#[derive(Copy, Clone, Debug, PartialEq, Serialize)]
pub enum GameResult {
    NaughtWin,
    CrossWin,
    Draw,
}

const BOARD_DIM: usize = 3;
pub const BOARD_SIZE: usize = BOARD_DIM * BOARD_DIM;

#[derive(Copy, Clone, Debug, PartialEq, Serialize)]
pub enum PlayerId {
    Naught,
    Cross,
}

#[derive(Debug)]
pub struct Board {
    state: [Option<PlayerId>; BOARD_SIZE],
    win_check_dirs: HashMap<i32, Vec<(i32, i32)>>,
}

#[allow(dead_code)] // TODO: Remove this.
impl Board {
    pub fn new(board: Option<Board>) -> Self {
        let state = board.map_or([None; BOARD_SIZE], |b| b.state);
        let win_check_dirs = HashMap::from([
            (0, vec![(1, 1), (1, 0), (0, 1)]),
            (1, vec![(1, 0)]),
            (2, vec![(1, 0), (1, -1)]),
            (3, vec![(0, 1)]),
            (6, vec![(0, 1)]),
        ]);

        Self {
            state,
            win_check_dirs,
        }
    }

    pub fn get_state(&self) -> [Option<PlayerId>; BOARD_SIZE] {
        self.state
    }

    fn coord_to_pos(&self, coord: (i32, i32)) -> i32 {
        coord.0 * BOARD_DIM as i32 + coord.1
    }

    fn pos_to_coord(self, pos: i32) -> (i32, i32) {
        (pos / BOARD_DIM as i32, pos % BOARD_DIM as i32)
    }

    pub fn reset(&mut self) {
        self.state = [None; BOARD_SIZE];
    }

    fn num_empty(&self) -> i32 {
        self.state.iter().filter(|&&x| x.is_none()).count() as i32
    }

    pub fn get_possible_next_moves(&mut self) -> Vec<usize> {
        let empty_cells: Vec<usize> = self
            .state
            .iter()
            .enumerate()
            .filter_map(|(i, &val)| if val.is_none() { Some(i) } else { None })
            .collect();

        empty_cells
    }

    fn is_legal(&self, pos: i32) -> bool {
        (0 <= pos && pos < BOARD_SIZE as i32) && (self.state[pos as usize].is_none())
    }

    pub fn make_move(
        &mut self,
        position: usize,
        side: PlayerId,
    ) -> ([Option<PlayerId>; BOARD_SIZE], Option<GameResult>) {
        if self.state[position].is_some() {
            panic!("Invalid move")
        }

        self.state[position] = Some(side);

        if self.check_win() {
            let winner = if side == PlayerId::Naught {
                GameResult::NaughtWin
            } else {
                GameResult::CrossWin
            };

            return (self.state, Some(winner));
        }

        if self.num_empty() == 0 {
            return (self.state, Some(GameResult::Draw));
        }

        (self.state, None)
    }

    fn apply_dir(&self, pos: i32, direction: (i32, i32)) -> i32 {
        let mut row = pos / 3;
        let mut col = pos % 3;
        row += direction.0;

        if !(0..=2).contains(&row) {
            return -1;
        }

        col += direction.1;

        if !(0..=2).contains(&col) {
            return -1;
        }

        row * 3 + col
    }

    fn check_win_in_dir(&self, pos: i32, direction: (i32, i32)) -> bool {
        let c = self.state[pos as usize];
        if c.is_none() {
            return false;
        }

        let p1 = self.apply_dir(pos, direction);
        let p2 = self.apply_dir(p1, direction);

        if p1 == -1 || p2 == -1 {
            return false;
        }

        if c == self.state[p1 as usize] && c == self.state[p2 as usize] {
            return true;
        }

        false
    }

    fn who_won(&self) -> Option<PlayerId> {
        for start_pos in &self.win_check_dirs {
            if self.state[*start_pos.0 as usize].is_some() {
                for direction in start_pos.1 {
                    let res = self.check_win_in_dir(*start_pos.0, *direction);
                    if res {
                        return self.state[*start_pos.0 as usize];
                    }
                }
            }
        }

        None
    }

    fn check_win(&self) -> bool {
        for start_pos in &self.win_check_dirs {
            if self.state[*start_pos.0 as usize].is_some() {
                for direction in start_pos.1 {
                    let res = self.check_win_in_dir(*start_pos.0, *direction);
                    if res {
                        return true;
                    }
                }
            }
        }

        false
    }

    fn state_to_char(&self, pos: i32) -> String {
        let ret = match self.state[pos as usize] {
            None => " ",
            Some(PlayerId::Naught) => "o",
            Some(PlayerId::Cross) => "x",
        };

        ret.to_string()
    }
}

impl ToString for Board {
    fn to_string(&self) -> String {
        let mut board_str = String::new();
        for i in 0..3 {
            board_str += &format!(
                "{}|{}|{}\n",
                self.state_to_char(i * 3),
                self.state_to_char(i * 3 + 1),
                self.state_to_char(i * 3 + 2)
            );
            if i != 2 {
                board_str += "-----\n"
            }
        }

        board_str += "\n";

        board_str
    }
}
