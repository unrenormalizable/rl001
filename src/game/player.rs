use crate::game::board::*;

pub trait Player {
    // TODO: Remove mut.
    fn new_game(&mut self, side: i32);

    fn make_move(&self, board: &mut Board) -> (GameResult, bool);

    fn final_result(&self, result: GameResult);
}
