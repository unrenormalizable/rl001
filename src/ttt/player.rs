use crate::ttt::board::*;

pub trait Player {
    fn new_game(&mut self, side: PlayerId);

    fn make_move(&mut self, board: &mut Board) -> Option<GameResult>;

    fn final_result(&self, result: GameResult);
}
