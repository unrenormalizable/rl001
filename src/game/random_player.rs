use crate::game::board::*;
use crate::game::player::*;
use std::option::*;

pub struct RandomPlayer {
    side: Option<i32>,
}

impl RandomPlayer {
    pub fn new() -> Self {
        Self { side: None }
    }
}

impl Player for RandomPlayer {
    fn new_game(&mut self, side: i32) {
        self.side = Some(side);
    }

    fn make_move(&self, board: &mut Board) -> (GameResult, bool) {
        let (_, res, finished) =
            board.make_move(board.random_empty_spot().unwrap(), self.side.unwrap());
        (res, finished)
    }

    fn final_result(&self, _result: GameResult) {}
}
