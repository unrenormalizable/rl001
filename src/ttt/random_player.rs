use super::board::*;
use super::player::*;
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use std::option::*;

pub struct RandomPlayer {
    side: Option<PlayerId>,
    rng: StdRng,
}

impl RandomPlayer {
    pub fn new(seed: Option<u64>) -> Self {
        let rng = match seed {
            Some(s) => StdRng::seed_from_u64(s),
            None => StdRng::from_rng(rand::thread_rng()).unwrap(),
        };

        Self { side: None, rng }
    }
}

impl Player for RandomPlayer {
    fn new_game(&mut self, side: PlayerId) {
        self.side = Some(side);
    }

    fn make_move(&mut self, board: &mut Board) -> Option<GameResult> {
        let moves = board.get_possible_next_moves();
        let pos = self.rng.gen_range(0..moves.len());

        let (_, res) = board.make_move(moves[pos], self.side.unwrap());

        res
    }

    fn final_result(&self, _result: GameResult) {}
}
