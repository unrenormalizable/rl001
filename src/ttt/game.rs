use super::board::*;
use super::player::*;
use serde::*;

#[derive(Serialize)]
pub struct GameState(PlayerId, Option<GameResult>, [Option<PlayerId>; BOARD_SIZE]);

pub fn play_game(
    board: &mut Board,
    player1: &mut dyn Player,
    player2: &mut dyn Player,
    callback: &mut impl FnMut(GameState),
) -> GameResult {
    player1.new_game(PlayerId::Naught);
    player2.new_game(PlayerId::Cross);
    board.reset();

    let final_result: GameResult;
    loop {
        let result = player1.make_move(board);
        callback(GameState(PlayerId::Naught, result, board.get_state()));

        match result {
            None => {}
            Some(r) => {
                final_result = r;
                break;
            }
        }

        let result = player2.make_move(board);
        callback(GameState(PlayerId::Cross, result, board.get_state()));
        match result {
            None => {}
            Some(r) => {
                final_result = r;
                break;
            }
        }
    }

    let _ = &player1.final_result(final_result);
    let _ = &player2.final_result(final_result);

    final_result
}

#[cfg(test)]
mod tests {
    use super::super::random_player::*;
    use super::*;

    #[test]
    fn test_games() {
        let mut board = Board::new(None);

        let mut p1 = RandomPlayer::new(Some(2718));
        let mut p2 = RandomPlayer::new(Some(2718));

        let mut draw_count = 0;
        let mut cross_count = 0;
        let mut naught_count = 0;

        let mut game_output = Vec::<GameState>::new();
        let mut callback = |gr| {
            game_output.push(gr);
        };

        for _ in 0..20 {
            let result = play_game(&mut board, &mut p1, &mut p2, &mut callback);
            match result {
                GameResult::CrossWin => cross_count += 1,
                GameResult::NaughtWin => naught_count += 1,
                GameResult::Draw => draw_count += 1,
            }
        }

        assert!(draw_count != 0);
        assert!(cross_count != 0);
        assert!(naught_count != 0);

        insta::assert_yaml_snapshot!(game_output);
    }
}
