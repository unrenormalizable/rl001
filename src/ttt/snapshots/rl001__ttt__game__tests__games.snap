---
source: src/ttt/game.rs
expression: game_output
---
- - Naught
  - ~
  - - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - Naught
    - Cross
    - ~
    - ~
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - Naught
    - Cross
    - ~
    - ~
    - ~
    - ~
    - Naught
- - Cross
  - ~
  - - Cross
    - ~
    - Naught
    - Cross
    - ~
    - ~
    - ~
    - ~
    - Naught
- - Naught
  - NaughtWin
  - - Cross
    - ~
    - Naught
    - Cross
    - ~
    - Naught
    - ~
    - ~
    - Naught
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Cross
    - ~
    - Naught
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Cross
    - Naught
    - Naught
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - Cross
    - ~
    - ~
    - Cross
    - Naught
    - Naught
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - Cross
    - ~
    - ~
    - Cross
    - Naught
    - Naught
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - Cross
    - Cross
    - ~
    - Cross
    - Naught
    - Naught
    - ~
    - ~
- - Naught
  - NaughtWin
  - - Naught
    - Cross
    - Cross
    - Naught
    - Cross
    - Naught
    - Naught
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - Cross
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - Naught
    - Cross
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - Naught
    - Cross
    - ~
    - ~
    - Naught
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - Naught
    - Cross
    - ~
    - Naught
    - Naught
    - Cross
    - ~
    - ~
- - Cross
  - ~
  - - Cross
    - Naught
    - Cross
    - ~
    - Naught
    - Naught
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - Cross
    - Naught
    - Cross
    - ~
    - Naught
    - Naught
    - Cross
    - ~
    - Naught
- - Cross
  - CrossWin
  - - Cross
    - Naught
    - Cross
    - Cross
    - Naught
    - Naught
    - Cross
    - ~
    - Naught
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - Cross
    - ~
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - Cross
    - Naught
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - Cross
    - Naught
    - Cross
    - Naught
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - Cross
    - Naught
    - Cross
    - Naught
    - Naught
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - Cross
    - Cross
    - Naught
    - Cross
    - Naught
    - Naught
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - Cross
    - Cross
    - Naught
    - Cross
    - Naught
    - Naught
    - ~
    - Naught
- - Cross
  - CrossWin
  - - ~
    - Cross
    - Cross
    - Naught
    - Cross
    - Naught
    - Naught
    - Cross
    - Naught
- - Naught
  - ~
  - - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - Naught
    - Cross
    - ~
    - ~
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - Naught
    - Cross
    - Naught
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - Naught
    - Cross
    - Naught
    - Cross
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - Naught
    - Cross
    - Naught
    - Cross
    - ~
    - ~
    - Naught
- - Cross
  - ~
  - - ~
    - ~
    - Naught
    - Cross
    - Naught
    - Cross
    - ~
    - Cross
    - Naught
- - Naught
  - ~
  - - ~
    - Naught
    - Naught
    - Cross
    - Naught
    - Cross
    - ~
    - Cross
    - Naught
- - Cross
  - ~
  - - ~
    - Naught
    - Naught
    - Cross
    - Naught
    - Cross
    - Cross
    - Cross
    - Naught
- - Naught
  - NaughtWin
  - - Naught
    - Naught
    - Naught
    - Cross
    - Naught
    - Cross
    - Cross
    - Cross
    - Naught
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - Cross
    - Naught
    - ~
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - ~
    - ~
    - Cross
    - Naught
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - Cross
    - ~
    - Cross
    - Naught
    - ~
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - Cross
    - ~
    - Cross
    - Naught
    - ~
    - Naught
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - Cross
    - ~
    - Cross
    - Naught
    - ~
    - Naught
    - Cross
    - ~
- - Naught
  - ~
  - - Naught
    - Cross
    - ~
    - Cross
    - Naught
    - Naught
    - Naught
    - Cross
    - ~
- - Cross
  - ~
  - - Naught
    - Cross
    - Cross
    - Cross
    - Naught
    - Naught
    - Naught
    - Cross
    - ~
- - Naught
  - NaughtWin
  - - Naught
    - Cross
    - Cross
    - Cross
    - Naught
    - Naught
    - Naught
    - Cross
    - Naught
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - Cross
    - Naught
    - ~
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - Naught
    - Cross
    - Naught
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - Cross
    - ~
    - Naught
    - Cross
    - Naught
    - ~
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - Cross
    - ~
    - Naught
    - Cross
    - Naught
    - ~
    - ~
    - ~
    - Naught
- - Cross
  - ~
  - - Cross
    - ~
    - Naught
    - Cross
    - Naught
    - ~
    - ~
    - Cross
    - Naught
- - Naught
  - NaughtWin
  - - Cross
    - ~
    - Naught
    - Cross
    - Naught
    - Naught
    - ~
    - Cross
    - Naught
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - ~
    - ~
    - Cross
    - ~
    - Naught
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Naught
    - Cross
    - ~
    - Naught
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Naught
    - Cross
    - ~
    - Naught
    - Cross
- - Naught
  - ~
  - - Naught
    - ~
    - ~
    - ~
    - Naught
    - Cross
    - ~
    - Naught
    - Cross
- - Cross
  - ~
  - - Naught
    - Cross
    - ~
    - ~
    - Naught
    - Cross
    - ~
    - Naught
    - Cross
- - Naught
  - ~
  - - Naught
    - Cross
    - ~
    - Naught
    - Naught
    - Cross
    - ~
    - Naught
    - Cross
- - Cross
  - CrossWin
  - - Naught
    - Cross
    - Cross
    - Naught
    - Naught
    - Cross
    - ~
    - Naught
    - Cross
- - Naught
  - ~
  - - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - Naught
    - ~
    - ~
    - ~
    - ~
    - Cross
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - Naught
    - ~
    - ~
    - ~
    - Cross
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - Naught
    - ~
    - Naught
    - ~
    - Cross
    - Cross
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - Naught
    - ~
    - Naught
    - Cross
    - Cross
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - Naught
    - ~
    - Naught
    - Cross
    - Cross
    - Cross
    - ~
    - Naught
- - Cross
  - CrossWin
  - - Naught
    - Naught
    - Cross
    - Naught
    - Cross
    - Cross
    - Cross
    - ~
    - Naught
- - Naught
  - ~
  - - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - ~
    - ~
    - ~
    - ~
    - Naught
    - Cross
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - ~
    - ~
    - Cross
    - ~
    - Naught
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - ~
    - Naught
    - Cross
    - ~
    - Naught
    - Cross
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - ~
    - Naught
    - Cross
    - Cross
    - Naught
    - Cross
    - ~
    - ~
- - Naught
  - NaughtWin
  - - Naught
    - ~
    - Naught
    - Cross
    - Cross
    - Naught
    - Cross
    - ~
    - Naught
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - Cross
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - Naught
    - ~
    - ~
    - Cross
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - Cross
    - Naught
    - Naught
    - ~
    - ~
    - Cross
    - ~
- - Naught
  - ~
  - - ~
    - Naught
    - Cross
    - Naught
    - Naught
    - ~
    - ~
    - Cross
    - ~
- - Cross
  - ~
  - - ~
    - Naught
    - Cross
    - Naught
    - Naught
    - Cross
    - ~
    - Cross
    - ~
- - Naught
  - ~
  - - ~
    - Naught
    - Cross
    - Naught
    - Naught
    - Cross
    - Naught
    - Cross
    - ~
- - Cross
  - ~
  - - Cross
    - Naught
    - Cross
    - Naught
    - Naught
    - Cross
    - Naught
    - Cross
    - ~
- - Naught
  - Draw
  - - Cross
    - Naught
    - Cross
    - Naught
    - Naught
    - Cross
    - Naught
    - Cross
    - Naught
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - Cross
- - Naught
  - ~
  - - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - Naught
    - ~
    - Cross
- - Cross
  - ~
  - - ~
    - ~
    - Naught
    - ~
    - Cross
    - ~
    - Naught
    - ~
    - Cross
- - Naught
  - ~
  - - Naught
    - ~
    - Naught
    - ~
    - Cross
    - ~
    - Naught
    - ~
    - Cross
- - Cross
  - ~
  - - Naught
    - Cross
    - Naught
    - ~
    - Cross
    - ~
    - Naught
    - ~
    - Cross
- - Naught
  - ~
  - - Naught
    - Cross
    - Naught
    - ~
    - Cross
    - Naught
    - Naught
    - ~
    - Cross
- - Cross
  - ~
  - - Naught
    - Cross
    - Naught
    - Cross
    - Cross
    - Naught
    - Naught
    - ~
    - Cross
- - Naught
  - Draw
  - - Naught
    - Cross
    - Naught
    - Cross
    - Cross
    - Naught
    - Naught
    - Naught
    - Cross
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - Cross
    - ~
    - Naught
- - Cross
  - ~
  - - ~
    - ~
    - Cross
    - ~
    - Naught
    - ~
    - Cross
    - ~
    - Naught
- - Naught
  - ~
  - - ~
    - ~
    - Cross
    - ~
    - Naught
    - Naught
    - Cross
    - ~
    - Naught
- - Cross
  - ~
  - - ~
    - ~
    - Cross
    - ~
    - Naught
    - Naught
    - Cross
    - Cross
    - Naught
- - Naught
  - NaughtWin
  - - ~
    - ~
    - Cross
    - Naught
    - Naught
    - Naught
    - Cross
    - Cross
    - Naught
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - Cross
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - Cross
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - Naught
- - Cross
  - ~
  - - Cross
    - ~
    - Cross
    - Naught
    - ~
    - ~
    - ~
    - ~
    - Naught
- - Naught
  - ~
  - - Cross
    - Naught
    - Cross
    - Naught
    - ~
    - ~
    - ~
    - ~
    - Naught
- - Cross
  - ~
  - - Cross
    - Naught
    - Cross
    - Naught
    - ~
    - ~
    - Cross
    - ~
    - Naught
- - Naught
  - ~
  - - Cross
    - Naught
    - Cross
    - Naught
    - Naught
    - ~
    - Cross
    - ~
    - Naught
- - Cross
  - ~
  - - Cross
    - Naught
    - Cross
    - Naught
    - Naught
    - Cross
    - Cross
    - ~
    - Naught
- - Naught
  - NaughtWin
  - - Cross
    - Naught
    - Cross
    - Naught
    - Naught
    - Cross
    - Cross
    - Naught
    - Naught
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - Cross
    - ~
- - Naught
  - ~
  - - Naught
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - Cross
    - ~
- - Cross
  - ~
  - - Naught
    - ~
    - ~
    - Naught
    - Cross
    - ~
    - ~
    - Cross
    - ~
- - Naught
  - ~
  - - Naught
    - ~
    - Naught
    - Naught
    - Cross
    - ~
    - ~
    - Cross
    - ~
- - Cross
  - CrossWin
  - - Naught
    - Cross
    - Naught
    - Naught
    - Cross
    - ~
    - ~
    - Cross
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - Cross
    - ~
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - Naught
    - ~
    - Naught
    - Cross
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - Naught
    - Cross
    - Naught
    - Cross
    - ~
    - ~
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - Naught
    - Cross
    - Naught
    - Cross
    - ~
    - ~
    - ~
    - Naught
- - Cross
  - ~
  - - Cross
    - Naught
    - Cross
    - Naught
    - Cross
    - ~
    - ~
    - ~
    - Naught
- - Naught
  - ~
  - - Cross
    - Naught
    - Cross
    - Naught
    - Cross
    - Naught
    - ~
    - ~
    - Naught
- - Cross
  - CrossWin
  - - Cross
    - Naught
    - Cross
    - Naught
    - Cross
    - Naught
    - Cross
    - ~
    - Naught
- - Naught
  - ~
  - - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - ~
    - ~
    - ~
    - ~
    - Naught
    - Cross
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - ~
    - ~
    - Cross
    - ~
    - Naught
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - ~
    - ~
    - Cross
    - Naught
    - Naught
    - Cross
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - Cross
    - ~
    - Cross
    - Naught
    - Naught
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - Cross
    - ~
    - Cross
    - Naught
    - Naught
    - Cross
    - Naught
    - ~
- - Cross
  - ~
  - - Naught
    - Cross
    - Cross
    - Cross
    - Naught
    - Naught
    - Cross
    - Naught
    - ~
- - Naught
  - NaughtWin
  - - Naught
    - Cross
    - Cross
    - Cross
    - Naught
    - Naught
    - Cross
    - Naught
    - Naught
- - Naught
  - ~
  - - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
    - Cross
    - ~
- - Naught
  - ~
  - - ~
    - Naught
    - ~
    - Naught
    - ~
    - ~
    - ~
    - Cross
    - ~
- - Cross
  - ~
  - - ~
    - Naught
    - ~
    - Naught
    - ~
    - ~
    - Cross
    - Cross
    - ~
- - Naught
  - ~
  - - ~
    - Naught
    - ~
    - Naught
    - Naught
    - ~
    - Cross
    - Cross
    - ~
- - Cross
  - ~
  - - ~
    - Naught
    - Cross
    - Naught
    - Naught
    - ~
    - Cross
    - Cross
    - ~
- - Naught
  - ~
  - - Naught
    - Naught
    - Cross
    - Naught
    - Naught
    - ~
    - Cross
    - Cross
    - ~
- - Cross
  - CrossWin
  - - Naught
    - Naught
    - Cross
    - Naught
    - Naught
    - ~
    - Cross
    - Cross
    - Cross
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - ~
    - ~
    - ~
    - Naught
    - Cross
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - ~
    - Naught
    - ~
    - Naught
    - Cross
    - ~
- - Cross
  - ~
  - - Cross
    - ~
    - ~
    - ~
    - Naught
    - ~
    - Naught
    - Cross
    - ~
- - Naught
  - ~
  - - Cross
    - ~
    - ~
    - ~
    - Naught
    - Naught
    - Naught
    - Cross
    - ~
- - Cross
  - ~
  - - Cross
    - ~
    - ~
    - ~
    - Naught
    - Naught
    - Naught
    - Cross
    - Cross
- - Naught
  - NaughtWin
  - - Cross
    - ~
    - Naught
    - ~
    - Naught
    - Naught
    - Naught
    - Cross
    - Cross
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - ~
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - ~
    - ~
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - ~
    - ~
    - ~
    - Naught
    - ~
    - Naught
    - Cross
    - ~
    - ~
- - Cross
  - ~
  - - ~
    - ~
    - Cross
    - Naught
    - ~
    - Naught
    - Cross
    - ~
    - ~
- - Naught
  - ~
  - - Naught
    - ~
    - Cross
    - Naught
    - ~
    - Naught
    - Cross
    - ~
    - ~
- - Cross
  - ~
  - - Naught
    - ~
    - Cross
    - Naught
    - ~
    - Naught
    - Cross
    - ~
    - Cross
- - Naught
  - ~
  - - Naught
    - ~
    - Cross
    - Naught
    - ~
    - Naught
    - Cross
    - Naught
    - Cross
- - Cross
  - CrossWin
  - - Naught
    - ~
    - Cross
    - Naught
    - Cross
    - Naught
    - Cross
    - Naught
    - Cross
